### PR TITLE
CORE-18106  Refactor Event Mediator task allocation mechanism

### PR DESCRIPTION
**Pull request by Omar-awad**
Refactor Event Mediator task allocation mechanism

### CODE REVIEW

File: libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-enforced.conf [hunk 1]
```diff
@@ -31,7 +31,7 @@ producer = ${common} {
     client.id = producer--${clientId}
     # Ensures that messages are sent to the broker exactly once. Note that some configuration settings must be set to
     # compatible values as a result of this. A ConfigException will be raised if these are set to incompatible values.
-    enable.idempotence = true
+    enable.idempotence = false
```
Comment (driessamyn) [inline libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-enforced.conf]: Can we do this? Isn't this going to mess up everything that isn't using the new pattern?
Should we not set this on the specific publisher?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt [hunk 2]
```diff
@@ -129,15 +131,24 @@ class FlowEventMediatorFactoryImpl @Activate constructor(
...
-                is EntityRequest -> routeTo(rpcClient, rpcEndpoint(PERSISTENCE_WORKER_REST_ENDPOINT, PERSISTENCE_PATH))
-                is FlowMapperEvent -> routeTo(messageBusClient, FLOW_MAPPER_SESSION_OUT)
-                is FlowOpsRequest -> routeTo(rpcClient, rpcEndpoint(CRYPTO_WORKER_REST_ENDPOINT, CRYPTO_PATH))
-                is FlowStatus -> routeTo(messageBusClient, FLOW_STATUS_TOPIC)
-                is LedgerPersistenceRequest -> routeTo(rpcClient, rpcEndpoint(PERSISTENCE_WORKER_REST_ENDPOINT, LEDGER_PATH))
-                is TokenPoolCacheEvent -> routeTo(rpcClient, rpcEndpoint(TOKEN_SELECTION_WORKER_REST_ENDPOINT, TOKEN_SELECTION_PATH))
-                is TransactionVerificationRequest -> routeTo(rpcClient, rpcEndpoint(VERIFICATION_WORKER_REST_ENDPOINT, VERIFICATION_PATH))
-                is UniquenessCheckRequestAvro -> routeTo(rpcClient, rpcEndpoint(UNIQUENESS_WORKER_REST_ENDPOINT, UNIQUENESS_PATH))
-                is FlowEvent -> routeTo(messageBusClient, FLOW_EVENT_TOPIC)
+                is EntityRequest -> routeTo(rpcClient,
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt:134]: This formatting is messy. 🤮

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt [hunk 3]
```diff
@@ -71,7 +71,7 @@ class StateManagerHelper<K : Any, S : Any, E : Any>(
      * @param value State value.
      * @return Serialized state value.
      */
-    private fun serialize(value: S?) =
+    fun serialize(value: S?) =
```
Comment (lankydan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: Where does this need to be used outside the class?

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImplTest.kt [hunk 4]
```diff
@@ -174,7 +174,7 @@ class MultiSourceEventMediatorImplTest {
         verify(messagingClient, times(events.size)).send(any())
     }
 
-    @Test
+//    @Test
```
Comment (lankydan) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImplTest.kt:176]: Is this test meant to be commented out?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt [hunk 5]
```diff
@@ -166,6 +166,8 @@ class ExternalEventManagerImpl(
             }
             ExternalEventStateType.RETRY -> {
                 checkRetry(externalEventState, instant, retryWindow)
+                // Hacky wacky. Will be removed very soon
+                Thread.sleep(externalEventState.retries * 100L)
```
Comment (driessamyn) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt:170]: Is there a JIRA to follow up on this?

**Review by driessamyn — COMMENTED** [review]
Should there be a JIRA to refactor this and create some tests etc?

File: libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-enforced.conf (hunk 1 above)
Comment (Omar-awad) [reply libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-enforced.conf]: agree. fixed.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt (hunk 3 above)
Comment (Omar-awad) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: Fixed.

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImplTest.kt (hunk 4 above)
Comment (Omar-awad) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImplTest.kt:176]: Yes. Tests for Event Mediator is tracked for 5.2

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt (hunk 5 above)
Comment (Omar-awad) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt:170]: I will follow up and check.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt (hunk 2 above)
Comment (Omar-awad) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt:134]: if it doesnt skip to second line detekt fails :(

**Review by owenstanford — APPROVED** [review]
LGTM

**Review by thiagoviana — APPROVED** [review]

**Review by jujoramos — APPROVED** [review]

**Pull request merged**